### PR TITLE
protocols/dcutr/example: Wait for relay to accept reservation request

### PR DESCRIPTION
**Pull request by mxinden**
# Description

When in listening mode, wait for the relay to accept our reservation
request. Only then can a client in dialing mode establish a relayed
connection to us via the relay.

## Links to any relevant issues

Might help with https://github.com/libp2p/rust-libp2p/issues/2621#issuecomment-1123549348.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by mxinden** [comment]
//CC @foemass

**Comment by mxinden** [comment]
@elenaf9 @thomaseizinger can one of you give this a review?

File: protocols/dcutr/examples/client.rs [hunk 1]
```diff
@@ -197,25 +197,41 @@ fn main() -> Result<(), Box<dyn Error>> {
...
                 SwarmEvent::Dialing { .. } => {}
                 SwarmEvent::ConnectionEstablished { .. } => {}
                 SwarmEvent::Behaviour(Event::Ping(_)) => {}
+                SwarmEvent::Behaviour(Event::Relay(client::Event::ReservationReqAccepted {
+                    ..
+                })) => {
+                    info!("Relay accepted our reservation request.");
+                    relay_accepted_reservation = true
+                }
                 SwarmEvent::Behaviour(Event::Relay(_)) => {}
```
Comment (elenaf9) [inline protocols/dcutr/examples/client.rs]: Maybe also catch the case of `client::Event::ReservationReqFailed` and return an error in that case. 
Shouldn't happen in practice but I'd rather have an explicit error than a non-terminating program. 
Alternatively, I think there is no valid case in which any other relay event should occur anyway/ So we could remove the match block here and let it fall to the blank panic below.

**Review by elenaf9 — APPROVED** [review]

File: protocols/dcutr/examples/client.rs [hunk 2]
```diff
@@ -197,25 +197,41 @@ fn main() -> Result<(), Box<dyn Error>> {
...
-                    info!("Observed address: {:?}", observed_addr);
-                    break;
+                    info!("Relay observes us under the address: {:?}", observed_addr);
+                    learned_observed_addr = true;
                 }
                 event => panic!("{:?}", event),
             }
+
+            if learned_observed_addr
+                && (matches!(opts.mode, Mode::Dial) || relay_accepted_reservation)
```
Comment (thomaseizinger) [inline protocols/dcutr/examples/client.rs]: Would this be easier to understand if we matched explicitly on `Mode` and returned `true` for dial and `relay_accepted_reservation` for listen?

**Review by thomaseizinger — APPROVED** [review]
LGTM

File: protocols/dcutr/examples/client.rs (hunk 2 above)
Comment (mxinden) [reply protocols/dcutr/examples/client.rs]: What do you think of https://github.com/libp2p/rust-libp2p/pull/2642/commits/75697c253d774294abdf2a00a40718f8a6d32493?

Comment (thomaseizinger) [reply protocols/dcutr/examples/client.rs]: Much better, thank you!

**Pull request merged**